crates/sys/src/lib.rs

```diff
@@ -29,12 +29,24 @@ mod inlines;
 )]
 mod bindings;
 
-/// Create a static C string.
+/// Create a static C string of type [`*const c_char`][core::ffi::c_char].
 /// Will automatically add a NUL terminator.
 #[macro_export]
 macro_rules! c_string {
     ($str:expr $(,)?) => {{
-        concat!($str, "\0").as_ptr() as *const core::ffi::c_char
+        ::core::concat!($str, "\0").as_ptr() as *const ::core::ffi::c_char
+    }};
+}
+
+/// Create a static C string of type [`&CStr`][`core::ffi::CStr`].
+/// Will automatically add a NUL terminator.
+#[macro_export]
+macro_rules! c_str {
+    ($str:expr $(,)?) => {{
+        match ::core::ffi::CStr::from_bytes_with_nul(::core::concat!($str, "\0").as_bytes()) {
```

Comment:
Since this is guaranteed to return a string with a NUL terminator, then we should be able to use [`CStr::from_bytes_with_nul`](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_bytes_with_nul_unchecked).

(I'm not 100% sure how macros behave with `#![deny(unsafe_code)]`. It may be worth switching to a generic const function if it does trigger the lint.)

Reply:
When used in `const` code (which is the expected use-case for the macro), panics should not end in the resulting code since the compiler should evaluate the resulting value, so I don't think that `panic` gere iş A problem.

As for `unsafe` in macro expansions, I am not sure how it interacts with `#![forbid(unsafe)] `, although proc-macro [`cstr`](https://docs.rs/cstr/0.2.11/cstr/) does generate `unsafe` call for this use-case.

By the way, it may be more rational to just re-export this macro from `cstr` crate rather than ad-hoc implement it here (specially, considering the fact that it does intermediate null-checking).

Reply:
Yeah. I'm not at all opposed to pulling in a crate if it provides better usability.

Currently the other benefit of [`CStr::from_bytes_with_nul_unchecked`]([CStr::from_bytes_with_nul](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.from_bytes_with_nul_unchecked) is that it has been `const` since 1.59.0.

Does `expect` work correctly in `const` evaluation? If so, that would be a better choice than `match`/`panic`.

Reply:
I've replaced the use with `cstr` crate which uses the `_unchecked` variant internally.
I've also made `c_string` delegate to it which has a minor memory overhead (since it now may statically store `CStr`'s meta) but it seems appropriate since we now get better compile-time checking.

In the future we may want to replace this with our own solution (or finally enjoy `c"foo"` syntax), but as for now it looks like the best solution.